src/coreComponents/constitutive/fluid/MultiFluidPVTPackageWrapper.hpp

```diff
@@ -155,13 +157,15 @@ class MultiFluidPVTPackageWrapperUpdate final : public MultiFluidBaseUpdate
                         real64 & dTotalDensity_dTemperature,
                         arraySlice1d< real64 > const & dTotalDensity_dGlobalCompFraction ) const override;
 
+  GEOSX_HOST_DEVICE
   GEOSX_FORCE_INLINE
   virtual void update( localIndex const k,
                        localIndex const q,
                        real64 const pressure,
                        real64 const temperature,
                        arraySlice1d< real64 const > const & composition ) const override
   {
+#ifndef __CUDACC__
```

Comment:
I'm a bit confused by all these function becoming `void` when not using `CUDA`.
Do we actually intend them to be void? If the `GPU` codes goes through them, will the results be OK?

Could it be possible that we simply remove them fore `CUDA` by _e.g._ inheriting a specialised class that would be compiled iff we do `CUDA` directly in the `CMakeLists.txt`. Or by putting the `#ifndef __CUDACC__` outside of the functions so they do not exist when not using `CUDA`?

Reply:
I feel like I should chime in since I proposed this. Let me briefly recap the discussion from https://github.com/GEOSX/GEOSX/pull/1313 (and sorry if I repeat points that are already clear).

  * The goal was, for a CUDA build (non-CUDA works as before), to allow different models to "opt-in" into different execution policies, such that models that aren't GPU-ready yet could still be run on the host.
  * To do this, each model declares (through a typedef) the policy it's ready to be run with, and the solver extracts the policy from the static type of the model (after constitutive dispatch) and uses it.
  * That by itself would work just fine. Problems only arise because we have used `virtual`/`override` keywords to enforce/document a common interface between these models. Even though we never use virtual dispatch, CUDA requires that all virtual/overriding functions share the same `host`/`device` designation.
  * Because **some** models want device code, **all** models now have to be marked `device`, which means `nvcc` will try to generate device code even for those functions that aren't device-ready, and will expectedly fail.
  * I saw two solutions to this: a) remove `virtual`/`override` that isn't mandatory; or b) make sure non-device-ready code is simply excluded from device compilation - but host version should still be generated - this is most easily achieved by just `#ifdef`-ing it out.
  * This function will never be called on GPU - the execution policy "opt-in" mechanism, if done correctly, makes sure of that. It becomes part of the class contract, if you will - the user **must** only use the policy the class presribes. So the function being a dummy is perfectly fine. 
  * Now, we *might* be able to `#ifdef` out the entire function definition as you suggest - I'm not sure if this will result in compiler or linker errors. Linker should be ok with this in theory, as the function is never called and thus doesn't need to be linked. Compiler should be ok in theory too - it does see the `override` declaration, so it knows the class claims implements all methods and can this be instantiated. But one can never be too sure with `nvcc`.
  * As for your other suggestion, I'm not sure I fully understood it - perhaps you can elaborate. I'll reiterate that we **want** this class to exist even in a CUDA build, and be usable on host, just not have compilation errors. And in the compilation model we use, any file that is part of the project is always compiled by `nvcc` in both host and device modes. The only way (that I'm aware of) to have those passes "sees" different code is through conditional compilation using `__CUDACC__` macro.

Reply:
> Linker should be ok with this in theory, as the function is never called and thus doesn't need to be linked.

I take it back. Since the function is virtual, its address needs to be put in the vtable regardless of whether it's going to be called or not. Not providing a definition (at least a dummy one) leads to an `unresolved reference to vtable` linker error.

Reply:
Hi @klevzoff so IIUC the root cause is that we use an abstract base class to enforce a class signature for a template pattern. Where we would like to use c++20's future `concept` but it's not ready yet. (Will it work with `__host__`/`__device__` stuff is another question).

My concern is that dummy function are still generated and therefore could be used by somebody not mastering completely the pattern (or simply someone tired 🤷 , that happens).
As such there is a risk that these functions are being indeed called, providing wrong results that may (I don't know) be complicated to detect, specially if we are unlucky enough to give this to an end-user.

I could propose two options to try to keep the abstract base class pattern while improving safety.

First immediate solution would be not to accept such that pattern:
```c++
__something__ void foo(...){
#ifndef __CUDACC__
   bar();
#endif
}
```
but just force the program to die in the catastrophic case where this function would still be called (I say catastrophic because if we end up calling this function with `CUDA` activated there is something really really wrong and the whole simulation may be corrected).
```c++
__something__ void foo(...){
#ifdef __CUDACC__
   die("NotImplementedError, you should not be there!");
#else
  bar();
#endif
}
```

An other, (but maybe less pattern protective I don't know) would be to have this `die` call directly in the base class (which may not be pure abstract anymore, depending on the need). Therefore a non overriden member function which would still be called would kill the simulation, and the derived class would require no dummy function.
I do not know the whole hierarchy implications, so that may that's not possbile.

Reply:
> First immediate solution would be not to accept that such pattern:
> ...
> but at least to force the program to die in the catastrophic case where this function would still be called. (I say catastrophic because if we end up calling this function with `CUDA` activated there is something really really wrong and the whole simulation may be flawed).

I understand now and completely agree - the function should throw/die when called in the unsupported execution space. Nice catch, and indeed extra protection like this is useful to have.

> An other, (but maybe less pattern protective I don't know) would be to have this `die` call directly in the base class (which may not be pure abstract anymore, depending on the need). Therefore a non overriden member function which would still be called would kill the simulation, and the derived class would require no dummy function.
> I do not know the whole hierarchy implications, so that may not be possbile.

So here the problem is we actually want to override `compute()` function in **all** derived classes - but some of those overrides want to be `__host__`-only and others `__host__ __device__`. The possibility to override in one execution space and have the other automatically call non-abstract base implementation would be great, but this scenario unfortunately isn't supported by `nvcc`: https://godbolt.org/z/cGP35E

Reply:
Here I changed the code as suggested by @TotoGaz  to force the problem to crash if this function is called when it `__CUDA_ACC_` is defined.